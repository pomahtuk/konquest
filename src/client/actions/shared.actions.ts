export enum SharedActionTypes {
  RESTORE_GAME = "RESTORE_GAME"
}

export interface RestoreGameAction {
  type: typeof SharedActionTypes.RESTORE_GAME;
  storedGame?: string;
}

export function restoreGame(storedGame?: string): RestoreGameAction {
  return { type: SharedActionTypes.RESTORE_GAME, storedGame };
}
