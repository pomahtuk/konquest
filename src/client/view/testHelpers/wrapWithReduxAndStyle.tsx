import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
