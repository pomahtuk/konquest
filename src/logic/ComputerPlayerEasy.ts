import ComputerPlayer, { ComputerPlayerType } from "./ComputerPlayer";

class ComputerPlayerEasy extends ComputerPlayer {
  protected minimumShips = 20;
  protected shipCountFactor = 2;

  constructor(name: string) {
    super(name, ComputerPlayerType.EASY);
  }
}

export default ComputerPlayerEasy;
