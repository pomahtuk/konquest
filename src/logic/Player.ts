export default class Player {
  public id: number;
  public screenName: string;

  public constructor(name: string) {
    // TODO: player actually have to be stored somewhere
    this.id = Math.floor(Math.random() * 1000);
    this.screenName = name;
  }
}

export interface PlayerMap {
  [key: string]: Player;
}
