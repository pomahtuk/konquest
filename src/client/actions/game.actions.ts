import Player, { PlayerTurn } from "../../logic/Player";

/*
 * action types
 */
export enum GameActionTypes {
  START_GAME = "START_GAME",
  SET_GAME_OPTIONS = "SET_GAME_OPTIONS",
  ADD_PLAYER_TURN = "ADD_PLAYER_TURN"
}

export interface StateGameOptions {
  fieldSize: number;
  neutralPlanetCount: number;
  players: Player[];
}

export interface StartGameAction {
  type: typeof GameActionTypes.START_GAME;
}

export interface SetGameOptionsAction {
  type: typeof GameActionTypes.SET_GAME_OPTIONS;
  gameOptions: StateGameOptions;
}

export interface AddPlayerTurnAction {
  type: typeof GameActionTypes.ADD_PLAYER_TURN;
  turnData: PlayerTurn;
}
/*
 * action creators
 */
export function startGame(): StartGameAction {
  return { type: GameActionTypes.START_GAME };
}

export function setGameOptions(gameOptions: StateGameOptions): SetGameOptionsAction {
  return { type: GameActionTypes.SET_GAME_OPTIONS, gameOptions };
}

export function addPlayerTurn(turnData: PlayerTurn): AddPlayerTurnAction {
  return { type: GameActionTypes.ADD_PLAYER_TURN, turnData };
}
