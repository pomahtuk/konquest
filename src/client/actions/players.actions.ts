import Player from "../../logic/Player";

export enum PlayerActionTypes {
  ADD_PLAYER = "ADD_PLAYER",
  REMOVE_PLAYER = "REMOVE_PLAYER",
  CLEAR_PLAYERS = "CLEAR_PLAYERS"
}

export interface AddPlayerAction {
  type: typeof PlayerActionTypes.ADD_PLAYER;
  player: Player;
}

export interface RemovePlayerAction {
  type: typeof PlayerActionTypes.REMOVE_PLAYER;
  player: Player;
}

export interface ClearPlayersAction {
  type: typeof PlayerActionTypes.CLEAR_PLAYERS;
}

// action creators
export function addPlayer(player: Player): AddPlayerAction {
  return { type: PlayerActionTypes.ADD_PLAYER, player };
}

export function removePlayer(player: Player): RemovePlayerAction {
  return { type: PlayerActionTypes.REMOVE_PLAYER, player };
}

export function clearPlayers(): ClearPlayersAction {
  return { type: PlayerActionTypes.CLEAR_PLAYERS };
}
