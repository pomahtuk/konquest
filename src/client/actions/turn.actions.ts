import { PlayerTurnOrder } from "../../logic/Player";
import Planet from "../../logic/Planet";

/*
 * action types
 */
export enum TurnActionTypes {
  SET_ORIGIN_PLANET = "SET_ORIGIN_PLANET",
  SET_DESTINATION_PLANET = "SET_DESTINATION_PLANET",
  ADD_PLAYER_TURN_ORDER = "ADD_PLAYER_TURN_ORDER",
  REMOVE_PLAYER_TURN_ORDER = "REMOVE_PLAYER_TURN_ORDER",
  CLEAR_TURN_DATA = "CLEAR_TURN_DATA",
  UNSET_PLANETS = "UNSET_PLANETS"
}

export interface SetPlanetAction {
  type: typeof TurnActionTypes.SET_ORIGIN_PLANET | typeof TurnActionTypes.SET_DESTINATION_PLANET;
  planet?: Planet;
}

export interface AddPlayerTurnOrderAction {
  type: typeof TurnActionTypes.ADD_PLAYER_TURN_ORDER;
  order: PlayerTurnOrder;
}

export interface ClearTurnDataAction {
  type: typeof TurnActionTypes.CLEAR_TURN_DATA;
}

export interface RemovePlayerTurnOrderAction {
  type: typeof TurnActionTypes.REMOVE_PLAYER_TURN_ORDER;
  index: number;
}

export interface UnsetPlanetsAction {
  type: typeof TurnActionTypes.UNSET_PLANETS;
}

/*
 * action creators
 */
export function setOriginPlanet(planet?: Planet): SetPlanetAction {
  return { type: TurnActionTypes.SET_ORIGIN_PLANET, planet };
}

export function setDestinationPlanet(planet?: Planet): SetPlanetAction {
  return { type: TurnActionTypes.SET_DESTINATION_PLANET, planet };
}

export function addPlayerTurnOrder(order: PlayerTurnOrder): AddPlayerTurnOrderAction {
  return { type: TurnActionTypes.ADD_PLAYER_TURN_ORDER, order };
}

export function removePlayerTurnOrder(index: number): RemovePlayerTurnOrderAction {
  return { type: TurnActionTypes.REMOVE_PLAYER_TURN_ORDER, index };
}

export function clearTurnData(): ClearTurnDataAction {
  return { type: TurnActionTypes.CLEAR_TURN_DATA };
}

export function unsetPlanets(): UnsetPlanetsAction {
  return { type: TurnActionTypes.UNSET_PLANETS };
}
