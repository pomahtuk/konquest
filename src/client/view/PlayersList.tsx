import React, { ReactElement, useContext } from "react";
import { useSelector, shallowEqual, useDispatch } from "react-redux";
import rootPlayersSelector from "../selectors/root.players.selector";
import Player from "../../logic/Player";
import { removePlayer } from "../actions/players.actions";
import { css } from "emotion";
import { ThemeContext } from "./themes/ThemeProvider";

const PlayersList = (): ReactElement | null => {
  const players: Player[] = useSelector(rootPlayersSelector, shallowEqual);
  const dispatch = useDispatch();
  const theme = useContext(ThemeContext);

  if (!players || players.length === 0) {
    return null;
  }

  const spanStyle = css`
    cursor: pointer;
    margin: 0 0 0 ${theme.units.medium};
    font-size: ${theme.fontSizes.smallest};
  `;

  return (
    <ul>
      {players.map((player) => (
        <li key={player.id}>
          <span>
            {player.isComputer ? "🤖" : "👨‍🚀"} {player.screenName}
          </span>
          <span className={spanStyle} onClick={() => dispatch(removePlayer(player))} data-testid="remove">
            ❌
          </span>
        </li>
      ))}
    </ul>
  );
};

export default PlayersList;
