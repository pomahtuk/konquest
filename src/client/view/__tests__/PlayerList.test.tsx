import React from "react";
import { render, fireEvent } from "@testing-library/react";

import wrapWithReduxAndStyle, { CurrentStore } from "../testHelpers/wrapWithReduxAndStyle";
import PlayersList from "../PlayersList";
import { clearPlayers, addPlayer } from "../../actions/players.actions";
import Player from "../../../logic/Player";
import ComputerPlayerEasy from "../../../logic/ComputerPlayerEasy";

const playerName = "test player";
const computerName = "test computer";

const playerSelector = `üë®‚ÄçüöÄ ${playerName}`;
const computerSelector = `ü§ñ ${computerName}`;

const testPlayer = new Player(playerName);
const testComputer = new ComputerPlayerEasy(computerName);

describe("<PlayersList />", (): void => {
  it("renders <PlayersList /> component without any players", (): void => {
    // cleanup store
    CurrentStore.dispatch(clearPlayers());
    const { container } = render(wrapWithReduxAndStyle(<PlayersList />));
    expect(container).toMatchSnapshot();
  });

  it("renders <PlayersList /> component without both type of players", (): void => {
    // cleanup store
    CurrentStore.dispatch(clearPlayers());
    CurrentStore.dispatch(addPlayer(testPlayer));
    CurrentStore.dispatch(addPlayer(testComputer));

    const { container, getByText } = render(wrapWithReduxAndStyle(<PlayersList />));
    expect(container).toMatchSnapshot();
    expect(getByText(playerSelector)).toBeDefined();
    expect(getByText(computerSelector)).toBeDefined();
  });

  it("<PlayersList /> can remove player", (): void => {
    // cleanup store
    CurrentStore.dispatch(clearPlayers());
    CurrentStore.dispatch(addPlayer(testPlayer));
    const { getByTestId, queryByText } = render(wrapWithReduxAndStyle(<PlayersList />));

    expect(queryByText(playerSelector)).not.toBe(null);
    //
    const removeButton = getByTestId("remove");
    fireEvent.click(removeButton);
    // verify that no player is defined
    expect(queryByText(playerSelector)).toBe(null);
  });
});
