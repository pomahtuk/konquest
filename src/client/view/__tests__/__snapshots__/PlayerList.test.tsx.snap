// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayersList /> renders <PlayersList /> component without any players 1`] = `<div />`;

exports[`<PlayersList /> renders <PlayersList /> component without both type of players 1`] = `
.emotion-0 {
  cursor: pointer;
  margin: 0 0 0 8px;
  font-size: 10px;
}

<div>
  <ul>
    <li>
      <span>
        👨‍🚀
         
        test player
      </span>
      <span
        class="emotion-0"
        data-testid="remove"
      >
        ❌
      </span>
    </li>
    <li>
      <span>
        🤖
         
        test computer
      </span>
      <span
        class="emotion-0"
        data-testid="remove"
      >
        ❌
      </span>
    </li>
  </ul>
</div>
`;
